### PR TITLE
Detect missing unique lowercase indexes for case insensitive unique validations

### PR DESCRIPTION
**Pull request by fatkodima**
Case insensitivity in uniqueness validations was skipped in this gem, but this is a very popular problem. 
I remember I was trapped by this at least once. And there is even [a related article from Richard Scheeman](https://schneems.com/2017/07/18/how-i-reduced-my-db-server-load-by-80/) with his investigation of the performance problems on production with his findings.   

It adds some ugly columns parsing, but I think we can deal with it.

### CODE REVIEW

File: lib/active_record_doctor/detectors/missing_unique_indexes.rb [hunk 1]
```diff
@@ -76,16 +89,8 @@ def has_ones_without_indexes # rubocop:disable Naming/PredicateName
...
-          validator.options[:unless].nil? &&
-          validator.options[:conditions].nil? &&
-
-          # In Rails 6, default option values are no longer explicitly set on
-          # options so if the key is absent we must fetch the default value
-          # ourselves. case_sensitive is the default in 4.2+ so it's safe to
-          # put true literally.
-          validator.options.fetch(:case_sensitive, true)
+      def conditional_validator?(validator)
+        (validator.options.keys & [:if, :unless, :conditions]).any?
```
Comment (gregnavis) [inline lib/active_record_doctor/detectors/missing_unique_indexes.rb:93]: I think `present?` would be slightly more appropriate here, as our intent is to test whether the intersection is non-empty, not whether it contains a truthy value.

File: lib/active_record_doctor/detectors/missing_unique_indexes.rb [hunk 2]
```diff
@@ -116,6 +123,26 @@ def ignore_columns
...
       end
+
+      def expression_indexes_unsupported?
+        (ActiveRecord::VERSION::STRING < "5.0") ||
+          # Active Record < 6 is unable to correctly parse expression indexes for MySQL.
+          (mysql? && ActiveRecord::VERSION::STRING < "6.0")
+      end
+
+      def index_columns(columns)
+        return columns unless columns.is_a?(String)
```
Comment (gregnavis) [inline lib/active_record_doctor/detectors/missing_unique_indexes.rb:134]: I'd like to make two points here:

1. Let's add a comment explaining why this is needed. What are types other than `String` that can be passed to the function.
2. Are there cases where this is an array and we'd need to perform the transformations below (those `gsub`s) on each of its elements?

File: lib/active_record_doctor/detectors/missing_unique_indexes.rb [hunk 3]
```diff
@@ -116,6 +123,26 @@ def ignore_columns
           column.gsub(" ", "")
         end
       end
+
+      def expression_indexes_unsupported?
+        (ActiveRecord::VERSION::STRING < "5.0") ||
+          # Active Record < 6 is unable to correctly parse expression indexes for MySQL.
+          (mysql? && ActiveRecord::VERSION::STRING < "6.0")
+      end
```
Comment (gregnavis) [inline lib/active_record_doctor/detectors/missing_unique_indexes.rb:131]: I suggest extracting this to `ActiveRecordDoctor.expression_indexes_unsupported?` and using it in this detector and the test suite.

File: test/active_record_doctor/detectors/missing_unique_indexes_test.rb [hunk 4]
```diff
@@ -179,12 +179,58 @@ def test_column_order_is_ignored
...
+    create_table(:users) do |t|
+      t.string :email
+
+      t.index :email, unique: true
+    end.create_model do
+      validates :email, uniqueness: { case_sensitive: false }
+    end
+
+    assert_problems(<<~OUTPUT)
+      add a unique index on users(lower(email)) - validating uniqueness in the model without an index can lead to duplicates
```
Comment (gregnavis) [inline test/active_record_doctor/detectors/missing_unique_indexes_test.rb:194]: I suggest a different error message when the complaint is about a missing **expression** index:

```
add a unique expression index on users(lower(email)) - validating case-insensitive uniqueness in the model without an expression index can lead to duplicates (a regular unique index is not enough)
```

File: test/active_record_doctor/detectors/missing_unique_indexes_test.rb [hunk 5]
```diff
@@ -179,12 +179,58 @@ def test_column_order_is_ignored
...
+
+    refute_problems
+  end
+
+  def test_case_insensitive_compound_lower_index_exists
+    skip("Expression indexes are not supported") if expression_indexes_unsupported?
+
+    create_table(:users) do |t|
+      t.string :email
+      t.integer :organization_id
```
Comment (gregnavis) [inline test/active_record_doctor/detectors/missing_unique_indexes_test.rb:220]: I think in this case we also should a unique index on `(organization_id, email)` to ensure only an **expression** index will satisfy the condition. Similarily, we need another test case with a non-unique expression index.

File: test/active_record_doctor/detectors/missing_unique_indexes_test.rb (hunk 4 above)
Comment (gregnavis) [reply test/active_record_doctor/detectors/missing_unique_indexes_test.rb:194]: Also, we need another test case that creates a **non-unique expression index** on the column.

The index must satisfy two conditions to satisfy the detector:

1. Unique
2. Expression

This gives four combinations in total:

1. No index at all (already covered by a previous test case).
2. Unique, no expression (the test case you added).
3. Non-unique, expression (the test case that's missing).
4. Unique, expression (added below).

**Review by gregnavis — CHANGES_REQUESTED** [review]

File: lib/active_record_doctor/detectors/missing_unique_indexes.rb (hunk 2 above)
Comment (fatkodima) [reply lib/active_record_doctor/detectors/missing_unique_indexes.rb]: 1. Added comment to the callsite
2. Seems like no. Rails returns columns as arrays only when they are "bare" columns. When one of them is an expression, the string is returned

File: lib/active_record_doctor/detectors/missing_unique_indexes.rb (hunk 3 above)
Comment (fatkodima) [reply lib/active_record_doctor/detectors/missing_unique_indexes.rb]: I extracted a separate module for this kind of utils. It feels not correct to me to pollute public `ActiveRecordDoctor` namespace with such helper functions. Let me know what you think.

**Comment by fatkodima** [comment]
Made suggested changes.

**Review by gregnavis — APPROVED** [review]

**Comment by gregnavis** [comment]
Merged! Thank you, @fatkodima.
